src/EditorFeatures/Core/InlineRename/InlineRenameSession.cs

```diff
@@ -651,26 +651,35 @@ public void Cancel()
     {
         _threadingContext.ThrowIfNotOnUIThread();
 
-        // This wait is safe.  We are not passing the async callback to DismissUIAndRollbackEditsAndEndRenameSessionAsync.
-        // So everything in that method will happen synchronously.
-        DismissUIAndRollbackEditsAndEndRenameSessionAsync(
-            RenameLogMessage.UserActionOutcome.Canceled, previewChanges: false).Wait();
+        DismissUIAndRollbackEditsAndEndRenameSession_MustBeCalledOnUIThread(
+            RenameLogMessage.UserActionOutcome.Canceled, previewChanges: false);
     }
 
-    private async Task DismissUIAndRollbackEditsAndEndRenameSessionAsync(
+    private void DismissUIAndRollbackEditsAndEndRenameSession_MustBeCalledOnUIThread(
+        RenameLogMessage.UserActionOutcome outcome,
+        bool previewChanges)
+    {
+        _threadingContext.ThrowIfNotOnUIThread();
+
+        DismissUIAndRollbackEditsAndEndRenameSession_MustBeCalledOnUIThread(
+            outcome, previewChanges, applyChangesOpt: null);
+    }
+
+    /// <summary>
+    /// Dismisses the UI, rolls back any edits, and ends the rename session.
+    /// </summary>
+    private void DismissUIAndRollbackEditsAndEndRenameSession_MustBeCalledOnUIThread(
         RenameLogMessage.UserActionOutcome outcome,
         bool previewChanges,
-        Func<Task> finalCommitAction = null)
+        Action applyChangesOpt)
```

Comment:
1. the applyChanges callback was made synchronous.
2. the function now required being on the UI therad.

these two things now make this entirely safe to call, without any sort of ugly .Wait from Cancel. 

Inside CommitAsync, we do all the actual async work we need, on the BG, prior to calling into this once we've switched back to the UI thread,